lib/active_merchant/billing/gateways/hi_pay.rb

```diff
@@ -0,0 +1,203 @@
+        post[:operation] = 'capture'
+        post[:currency] = (options[:currency] || currency(money))
+
+        commit('capture', post, { transaction_reference: authorization })
+      end
+
+      def store(payment_method, options = {})
+        tokenize(payment_method, options.merge({ multiuse: '1' }))
+      end
+
+      def scrub(transcrip)
+        # code
+      end
+
+      private
+
+      def add_product_data(post, options)
+        post[:orderid] = options[:order_id] if options[:order_id]
+        post[:description] = options[:description]
+
+        # optional
+        # basket=> refer https://support.hipay.com/hc/fr/articles/115001660469-Payment-Gateway-Shopping-cart-management
+        post[:basket] = options[:basket] if options[:basket]
+        post[:long_description] = options[:long_description] if options[:long_description]
+        post[:shipping] = options[:shipping_fee] if options[:shipping_fee]
+        post[:tax] = options[:tax] if options[:tax]
+        post[:tax_rate] = options[:tax_rate] if options[:tax_rate]
+      end
+
+      def add_invoice(post, money, options)
+        post[:currency] = (options[:currency] || currency(money))
+        post[:amount] = money
+      end
+
+      def add_credit_card(post, credit_card)
+        post[:card_number] = credit_card.number
+        post[:card_expiry_month] = credit_card.month
+        post[:card_expiry_year] = credit_card.year
+        post[:card_holder] = credit_card.name
+        post[:cvc] = credit_card.verification_value
+      end
+
+      def add_address(post, options)
+        if options[:billing_address]
+          post[:streetaddress] = options[:billing_address][:address1] if options[:billing_address][:address1]
+          post[:streetaddress2] = options[:billing_address][:address2] if options[:billing_address][:address2]
+          post[:city] = options[:billing_address][:city] if options[:billing_address][:city]
+          post[:recipient_info] = options[:billing_address][:company] if options[:billing_address][:company]
+          post[:state] = options[:billing_address][:state] if options[:billing_address][:state]
+          post[:country] = options[:billing_address][:country] if options[:billing_address][:country]
+          post[:zipcode] = options[:billing_address][:zip] if options[:billing_address][:zip]
+          post[:country] = options[:billing_address][:country] if options[:billing_address][:country]
+          post[:phone] = options[:billing_address][:phone] if options[:billing_address][:phone]
+        end
+      end
+
+      def tokenize(payment_method, options = {})
+        post = {}
+        add_credit_card(post, payment_method)
+        post[:multi_use] = options[:multiuse] ? '1' : '0'
+        post[:generate_request_id] = '0'
+        commit('store', post, options)
+      end
+
+      def parse(body)
+        return {} if body.blank?
+
+        JSON.parse(body)
+      end
+
+      def commit(action, post, options = {})
+        begin
```

Comment:
💬 Code Style / Comment:

Just to prevent a double call to `parse`:

```ruby
raw_response = begin
                 ssl_post(url(action, options), post_data(post), request_headers)
               rescue StandardError => e
                 e.response.body
               end

response = parse(raw_response)

```

Reply:
Thanks